### ISSUE TITLE
Clicking cropping controls in Volume Rendering module triggers unrequested actions

### ISSUE THREAD

**Issue by DIV-on-github** [Slicer/Slicer#6126]
### Description of issue
"in the "Crop" settings within the Display panel, clicking the Display ROI (eye) icon to display a ROI in the viewer automatically causes the Enable box to be ticked (enabling cropping according to that ROI) — despite the fact that there is a separate control provided.
[On the other hand, clicking the Display ROI (eye) icon to hide a ROI in the viewer does not affect the Enable setting. Which is as it should be, in my view.]"
_Originally posted by @DIV-on-github in https://github.com/Slicer/Slicer/issues/5290#issuecomment-1018213289_

This would be relevant when it is desired to continue displaying a cropped/uncropped volume whilst simultaneously _displaying_ additional ROI's as an 'overlay' (without _enabling_ them).  
Currently to do this is inconvenient, resulting in extra user actions to revert the undesirably activated ROI's.  

#### Analogy
An analogy could be clicking Print in a word processor.  I expect that this would automatically initiate a few 'essential' tasks, such as repagination and updating cross-references, especially when such tasks are not normally controlled manually.  But I wouldn't expect that it would trigger 'optional' tasks such as a spelling check (even though it could be argued that spelling should be checked before printing), because there's a separate control to check spelling, and the user can choose for themselves when (or if) to run this.  

### Opposing viewpoints
"FWIW, I find that behavior very useful - I almost always want to enable and display in one click, adjust the ROI, then make the ROI invisible."
_Originally posted by @pieper in https://github.com/Slicer/Slicer/issues/5290#issuecomment-1018479789_

"[...] I find it really convenient. When I enable cropping the first time I always need to display the ROI. Hiding the ROI must not disable cropping, because the ROI very often needs to be hidden (to declutter the view), while the cropping still should be applied. So, the current behavior is exactly what I need. "
_Originally posted by @lassoan in https://github.com/Slicer/Slicer/issues/5290#issuecomment-1018764188_

### Possible compromise solution
* Maintain the current behaviour of 'tacitly' enabling the ROI upon clicking `Display ROI` _when that ROI did not previously exist_.
* Amend to **not** 'tacitly' enabling the ROI upon clicking `Display ROI` _when that ROI already exists_ and can therefore be presumed to have already been configured as desired.

### Reasonable behaviour
_When 3D Slicer is opened and a volume is first loaded_, no ROI's are defined.  (The initially hidden _Inputs_ panel shows `ROI` set to "None".)  Then when clicking either `Enable` or `Display ROI` a ROI called "Volume rendering ROI" is automatically created (and selected in the _Inputs_ panel), which matches the extents of the volume.  Even though there are separate controls for creating a ROI, it does seem reasonable to create such a ROI under these circumstances.  (In fact, I consider it is more like 'efficient' programming to act as though the default ROI was available as soon as the volume was loaded.)

### Undesirable behaviour
_Later, when one or more ROI's have been defined_, but where `ROI` has been manually set to "None" in the _Inputs_ panel, when clicking either `Enable` or `Display ROI` a ROI called "Volume rendering ROI" is automatically created (and selected in the _Inputs_ panel), which matches the extents of the volume.  
In this way the user might unexpectedly create multiple (perhaps identical) ROI's.  

### Suggested solution
* Maintain the current behaviour of 'tacitly' creating a new ROI upon clicking `Display ROI` _when no ROI previously existed_.
* Amend to **not** 'tacitly' creating a new ROI upon clicking `Display ROI` _when any ROI already exists_.  Instead either (i) do nothing [because the user has manually selected "None", and we assume that they know what they're doing], or (ii) just act on the first available ROI in the list, or (iii) produce an error message or dialogue box.  

## Environment
- Slicer version: Slicer-4.13.0-2022-01-19
- Operating system: Windows 10

**Comment by lassoan** [comment]
I agree that conceptually it would be cleaner to not couple ROI enabling and visibility. However, we strive to "make simple things simple and complex things possible". This dictates that number of clicks should be minimized for the most commonly needed workflows. If unusual workflows are still feasible (with potentially a few extra clicks) then we achieved our goals. 

Please describe your clinical end goal (e.g., "I need to crop PET and CT volumes to the same ROI because...") to see if there are commonly needed workflows that are made unreasonably complicated by the current behavior. Also describe if you have any special workflow that is made impossible by the current implementation.

We could add an application settings if there is unresolvable difference between preferences of different user groups, but we would only do this if there are enough users in both groups There are things that we can do for one single user, too. For example we can add a new property to the volume rendering Qt widget that enables/disables auto-show (and you can change that in your application startup script).

**Comment by DIV-on-github** [comment]
## Sub-issue (a): undesirable enabling of cropping
> However, we strive to "make simple things simple and complex things possible". This dictates that number of clicks should be minimized for the most commonly needed workflows. If unusual workflows are still feasible (with potentially a few extra clicks) then we achieved our goals.

I agree that this makes sense.  So far it's 2:1 against changing....

> For example we can add a new property to the volume rendering Qt widget that enables/disables auto-show (and you can change that in your application startup script).

That would be a reasonable approach. 

> Also describe if you have any special workflow that is made impossible by the current implementation.

The current behaviour doesn't make anything "impossible", it's more accurately described as a minor inconvenience/irritation (to me).  

## Sub-issue (b): automated creation of ROI's
Could you please comment on this too?

**Comment by lassoan** [comment]
Creating a ROI automatically when user requests using a ROI is the best behaviour I can think of. The alternative options you have considered:
- (i) do nothing [because the user has manually selected "None", and we assume that they know what they're doing] -> users very often don't know what they are doing (just clicking around to experiment) or they may receive a scene where the ROI is set to None, or they delete the ROI in the Data module and have no idea that it would make it necessary to open the Inputs section and create a ROI (they may not even know that a ROI can be manually selected there). This would bring us lots of error reports that "volume rendering cropping does not work". 
-  (ii) just act on the first available ROI in the list -> most likely the ROI there is used for some other purpose
-  (iii) produce an error message or dialogue box - > this would require creating a hidden flag in the volume rendering display node that would store that a valid ROI has been ever selected. It would need to be hidden because we would not want to add an "Auto-create ROI if needed" checkbox to make the module's GUI more complicated. Such hidden states confuse the users, they may just think the application works incorrectly/inconsistently. We could show a popup to users all the time (not just after the ROI was set to None once) with a "don't show again" checkbox. However, when your show users unexpected popups while they are trying to do something, they immediately dismiss it after reading the first 1-2 words. So, all the popup would achieve is slightly annoying the users. 

Any frustration by any user indicates that something is not right, that the current implementation could or should be improved. So, I'm sure that we could do something about your pain points that may help all users. I just don't know exactly what. Maybe the reason of your frustration is that the current behavior is complex (clicking a button has side effects), so you are not sure what is happening exactly, so if you do want to know it then you need to spend time with experimentation. Describing all what the enable and display ROI buttons do in tooltips and/or documentation could help. Maybe making somehow the ROI node selector appear closely to the enable and display ROI checkboxes could help, too (we would need some good ideas how to make it available there without cluttering the GUI). Making the behavior simpler by disabling all convenience mechanisms (ROI auto-create and auto-show) would be relatively easy to implement, too, but we would need to make sure that the options are discoverable (if people cannot find a feature then it is not useful) yet do not clutter the GUI.

**Comment by DIV-on-github** [comment]
## Sub-issue (a): undesirable enabling of cropping
For completeness another idea I had was to contextually change the text label next to the 'eye' icon to read either
* Display & enable ROI
* Hide ROI

Personally I am not keen on that idea myself, but maybe it can spark some better suggestions.  

## Sub-issue (b): automated creation of ROI's
Here I thought it might make sense to 'deactivate' (grey out) the "Display ROI" and "Enable" controls if ROI is set to None.  But that would be the status when first launching 3D Slicer, and it'd be inconvenient to be forced to create a ROI 'manually' (as you've alluded to).  However, this suggests a new GUI configuration for the _DIsplay_ panel in the **Volume Rendering** module:  

-- -- --
▼ Display
Preset:  [Select a Preset]
Shift:  -------------------------0---------------
Crop: □ Enable ROI  ◎ Display ROI  <**New ROI**>  <▩ Fit to Volume>
-- -- --

In other words, adding an explicit "**New ROI**" button.  
The interface becomes slightly more crowded, but I think there's space.  
If necessary it can be shortened, as in 
Crop: □ Enable ◎ Display <**New**>  <▩ Fit to Volume>
and use tooltips to clarify.  

Anyway, by doing this it becomes sensible to grey out the 'display' and 'enable' buttons when no ROI is selected.  
Also, I think users would expect that upon creating a new ROI it would certainly be displayed, and they would be unsurprised if it were also enabled.  

At this moment there may not be a big demand for such changes, so I understand that it's not the highest priority (and thanks for spending time to consider it).

**Comment by pieper** [comment]
Adding an explicit `New ROI` button seems reasonable to me.
